### PR TITLE
[manila-csi-plugin] Replaced ChooseExportLocation with FindExportLocation

### PR DESCRIPTION
**Pull request by gman0**
**What this PR does / why we need it**:

This PR replaces `ChooseExportLocation` with `FindExportLocation` and adds a unit test for this function.

`FindExportLocation` retains the behavior of its predecessor, but adds the ability to further constrain the result by using a filter function that's supplied to `FindExportLocation` as an argument.

We'll need this for the NFS share adapter (at least), where we'll need to filter out unsuitable export locations (e.g. those that are not from a certain subnet).

xref https://github.com/kubernetes/cloud-provider-openstack/issues/1136

**Release note**:

```release-note
NONE
```

### CODE REVIEW

**Comment by gman0** [comment]
PTAL @tsmetana @tombarron @gouthampacha @ramineni

**Comment by gman0** [comment]
FYI @JPRbrs there will be one or (at most) two more PRs till we get a fix for https://github.com/kubernetes/cloud-provider-openstack/issues/1136 in

**Comment by JPRbrs** [comment]
Thanks Robert!

File: pkg/csi/manila/util/exportlocation.go [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+)
+
+// Predicate type for filtering out export locations.
+// It's supplied with the index into the export location slice that's being scanned.
+// Should return (true, nil) when search conditions are satisfied,
+// and (false, nil) when they are not.
+type ExportLocationPredicate func(index int) (match bool, err error)
+
+// Predicate matching any export location
+func AnyExportLocation(int) (bool, error) { return true, nil }
```
Comment (ramineni) [inline pkg/csi/manila/util/exportlocation.go:30]: Are any other predicates supported now?

File: pkg/csi/manila/util/exportlocation.go [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+// Searches for an export location.
+// Returns index of an export location from the `locs` slice that satisfies following rules:
+// 1. Location is not admin-only and is not empty
+// 2. Location satisfies ExportLocationPredicate
+// Search is biased:
+// 1. Location.Preferred == true is preferred over Location.Preferred == false
+// 2. Locations with lower index are preferred over those with higher index
+func FindExportLocation(locs []shares.ExportLocation, pred ExportLocationPredicate) (index int, err error) {
+	const invalidIdx = -1
+	firstMatchNotPreferred := invalidIdx
```
Comment (ramineni) [inline pkg/csi/manila/util/exportlocation.go:41]: confused, why is the name firstMatchNotPreferred , I guess this stores, first preferred match on contrary

File: pkg/csi/manila/util/exportlocation.go (hunk 1 above)
Comment (gman0) [reply pkg/csi/manila/util/exportlocation.go:30]: Not yet, we're planning two more for now: 1) exact match, and 2) subnet match.

File: pkg/csi/manila/util/exportlocation.go (hunk 2 above)
Comment (gman0) [reply pkg/csi/manila/util/exportlocation.go:41]: A preferred location is `locs[i].Preferred==true`, which makes the function return early:

https://github.com/gman0/cloud-provider-openstack/blob/26613136736920cd667fa28fdc56d8b103cd19fc/pkg/csi/manila/util/exportlocation.go#L51-L55

When not preferred (`Preferred==false`), we keep going through the list, searching for a preferred location. If none is found, we return the first matching "not preferred" location.

**Comment by gman0** [comment]
@ramineni this is in preparation for an upcoming PR that will fix the issue

**Comment by gman0** [comment]
@ramineni I see [gouthampacha's "approve"](https://github.com/kubernetes/cloud-provider-openstack/pull/1157#issuecomment-673720413) didn't trigger merge, even though he is an approver for `pkg/csi/manila`, and this PR already had a LGTM label at that time. Any idea why this might be?

File: pkg/csi/manila/util/exportlocation.go (hunk 2 above)
Comment (ramineni) [reply pkg/csi/manila/util/exportlocation.go:41]: Ah, got it, you are returning if preferred.

**Comment by ramineni** [comment]
@gman0 This PR is merged.

**Pull request merged**